app/src/server/api/routers/blackmarket.ts

```diff
@@ -405,49 +405,144 @@ export const blackMarketRouter = createTRPCRouter({
       }
     }),
   rerollElement: protectedProcedure
+    .input(z.object({ elementType: z.enum(["primary", "secondary"]) }))
     .output(baseServerResponse)
-    .mutation(async ({ ctx }) => {
+    .mutation(async ({ ctx, input }) => {
       // Fetch
       const user = await fetchUser(ctx.drizzle, ctx.userId);
       // Guard
       if (user.reputationPoints < COST_REROLL_ELEMENT) {
         return errorResponse("Not enough reputation points");
       }
+      
       // Get the updated elements
       const rankId = UserRanks.findIndex((r) => r === user.rank);
-      if (rankId >= 1) {
-        const available = BasicElementName.filter((e) => e !== user.primaryElement);
-        user.primaryElement = getRandomElement(available) ?? null;
+      const changes: string[] = [];
+      
+      // Get rolled elements from actionLog
+      const rolledElementsData = await getRolledElements(ctx.drizzle, ctx.userId, input.elementType);
+      
+      // Ensure current elements are in actionLog if not already tracked
+      if (user.primaryElement && !rolledElementsData.primary.includes(user.primaryElement)) {
+        await addElementRoll(ctx.drizzle, ctx.userId, "primary", user.primaryElement);
+        rolledElementsData.primary.push(user.primaryElement);
+      }
+      if (user.secondaryElement && !rolledElementsData.secondary.includes(user.secondaryElement)) {
+        await addElementRoll(ctx.drizzle, ctx.userId, "secondary", user.secondaryElement);
+        rolledElementsData.secondary.push(user.secondaryElement);
       }
-      if (user.secondaryElement) {
-        const available = BasicElementName.filter(
-          (e) => ![user.primaryElement, user.secondaryElement].includes(e),
+      
+      // Helper function to filter valid rolled elements
+      const filterValidRolledElements = (elements: string[]): typeof ElementNames[number][] => {
+        return elements.filter(
+          (e): e is typeof ElementNames[number] => ElementNames.includes(e as typeof ElementNames[number])
         );
-        user.secondaryElement = getRandomElement(available) ?? null;
+      };
+      
+      // Extract common reroll logic
+      const rerollElementType = async (
```

Comment:
Why does this convenience function need to be defined as a closure, when the others are outside the endpoint?